pkg/logql/sketch/topk.go

```diff
@@ -413,25 +413,48 @@ func (t *Topk) Topk() []TopKResult {
 
 func (t *Topk) TopkForGroupingKey(groupingKey string) TopKResult {
 	n := t.max
-	if len(*t.heaps[groupingKey]) < t.max {
-		n = len(*t.heaps[groupingKey])
+	if len(*t.Heaps[groupingKey]) < t.max {
+		n = len(*t.Heaps[groupingKey])
 	}
 	res := TopKResult{
 		groupingKey: "",
-		Result:      make([]element, 0, len(*t.heaps[groupingKey])),
+		Result:      make([]element, 0, len(*t.Heaps[groupingKey])),
 	}
 
-	for _, e := range *t.heaps[groupingKey] {
+	for _, e := range *t.Heaps[groupingKey] {
 		res.Result = append(res.Result, element{
-			Event: e.event,
-			Count: int64(t.sketch.Count(e.event)),
+			Event: e.Event,
+			Count: int64(t.sketch.Count(e.Event)),
 		})
 	}
 	sort.Sort(res)
 	res.Result = res.Result[:n]
 	return res
 }
 
+func (t *Topk) GroupingKeys() []string {
+	keys := make([]string, 0, len(t.Heaps))
+	for key := range t.Heaps {
+		keys = append(keys, key)
+	}
+	return keys
+}
+
+// DistinctEventsForGroupingKey returns a list of distinct events observed.
+// Returns nil if group was not found.
+func (t *Topk) DistinctEventsForGroupingKey(groupingKey string) []string {
+	group, ok := t.Heaps[groupingKey]
+	if !ok {
+		return nil
+	}
+
+	events := make([]string, 0, len(*group))
+	for _, node := range *group {
+		events = append(events, node.Event)
+	}
+	return events
+}
```

Comment:
The function names here are a bit confusing. Maybe I'm just missing what we would use these functions for other than in tests?

Couldn't you just use the function that returns the topk for the grouping key, and then iterate over the topk just selecting the event names?

Reply:
`topk` is not including all events. This is just a convenient way to access all observed events without exposing all internals.

Reply:
This is essentially just doing what `TopkForGroupingKey` is doing but not returning the counts for each event. The heap already is the entire list of events we're tracking for the grouping key. We don't track _all_ observed events.